### PR TITLE
SNOW-352846 Authentication - OAuth

### PR DESCRIPTION
**Pull request by sfc-gh-alhuang**

### CODE REVIEW

File: src/main/java/net/snowflake/ingest/connection/JWTManager.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2012-2017 Snowflake Computing Inc. All rights reserved.
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/JWTManager.java:2]: Header should be 2023

**Review by sfc-gh-tjones — COMMENTED** [review]
A general comment: it might be good to break this into a few PRs that build on each other. Right now the change is over 900 lines.

File: src/main/java/net/snowflake/ingest/connection/JWTManager.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+/**
+ * This class manages creating and automatically renewing the JWT token
+ *
+ * @author obabarinsa
+ * @since 1.8
+ */
+final class JWTManager extends SecurityManager {
+
+  // the token lifetime is 59 minutes
+  private static final float LIFETIME = 59;
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/JWTManager.java:34]: Let's add a time unit here: `LIFETIME_IN_MINUTES`

File: src/main/java/net/snowflake/ingest/connection/JWTManager.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+  // the renewal time is 54 minutes
+  private static final int RENEWAL_INTERVAL = 54;
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/JWTManager.java:37]: Same here: `RENEWAL_INTERVAL_IN_MINUTES`

File: src/main/java/net/snowflake/ingest/connection/OAuthCredential.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package net.snowflake.ingest.connection;
+
+import java.util.Base64;
+
+public class OAuthCredential {
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthCredential.java:5]: Nit, docstring for public class.

File: src/main/java/net/snowflake/ingest/connection/OAuthCredential.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  private static final String BASIC_AUTH_PREFIX = "Basic ";
+  private final String authHeader;
+  private final String clientId;
+  private final String clientSecret;
+  private String accessToken;
+  private final String refreshToken;
+
+  public OAuthCredential(String clientId, String clientSecret, String refreshToken) {
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthCredential.java:13]: Nit, docstring for public constructor.

File: src/main/java/net/snowflake/ingest/connection/OAuthCredential.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package net.snowflake.ingest.connection;
+
+import java.util.Base64;
+
+public class OAuthCredential {
+  private static final String BASIC_AUTH_PREFIX = "Basic ";
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthCredential.java:6]: Is this for the header? Maybe this should be `BASIC_AUTH_HEADER_PREFIX`

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2012-2017 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.ingest.connection;
+
+import static net.snowflake.ingest.utils.Constants.*;
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthManager.java:7]: Nit, no wildcard imports.

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java [hunk 8]
```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2012-2017 Snowflake Computing Inc. All rights reserved.
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthManager.java:2]: This should be 2023.

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+import net.snowflake.client.jdbc.internal.google.api.client.http.HttpStatusCodes;
+import net.snowflake.client.jdbc.internal.google.gson.JsonObject;
+import net.snowflake.client.jdbc.internal.google.gson.JsonParser;
+import net.snowflake.ingest.utils.*;
+
+/**
+ * This class manages creating and automatically renewing the JWT token
+ *
+ * @author obabarinsa
+ * @since 1.8
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthManager.java:38]: This needs to be changed: this is for renewing the OAuth token and that is not the author :)

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+ */
+final class OAuthManager extends SecurityManager {
+
+  // Default update threshold, a floating-point value representing the ratio between the expiration
+  // time of an access
+  // token and the time needed to update it. It must be a value greater than 0 and less than or
+  // equal to 1.
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthManager.java:45]: Nit, this seems improperly formatted.

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+
+  private static final String ACCESS_TOKEN = "access_token";
+  private static final String REFRESH_TOKEN = "refresh_token";
+  private static final String EXPIRES_IN = "expires_in";
+  private final double updateThresholdRatio;
+
+  private final AtomicReference<OAuthCredential> oAuthCredential;
+
+  // Did we fail to regenerate our token at some point?
+  private final AtomicBoolean refreshFailed;
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthManager.java:60]: We might want to have a retry policy as there could be a transient network hiccup that blocks us from obtaining a new token.

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+  private final CloseableHttpClient httpClient;
+
+  // Thread factory for daemon threads so that application can shutdown
+  final ThreadFactory tf = ThreadFactoryUtil.poolThreadFactory(getClass().getSimpleName(), true);
+
+  // the thread we use for refreshing access token
+  private final ScheduledExecutorService tokenRefresher = Executors.newScheduledThreadPool(1, tf);
+
+  // Reference to the Telemetry Service in the client
+  private final TelemetryService telemetryService;
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthManager.java:73]: Looks like these are in the JWT token manager as well - can we move them to a shared base class so we don't need to duplicate these?

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+  OAuthManager(
+      String accountName,
+      String username,
+      OAuthCredential oAuthCredential,
+      URI tokenRequestURI,
+      double updateThresholdRatio,
+      TelemetryService telemetryService) {
+    super(accountName, username);
+    // if any of our arguments are null, throw an exception
+    if (oAuthCredential == null) {
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthManager.java:119]: Looks like we only check `oAuthCredential` and not the rest of the args?

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+    for (int retries = 0; retries < MAX_REFRESH_TOKEN_RETRY; retries++) {
+      String respBodyString = null;
+      try (CloseableHttpResponse httpResponse = httpClient.execute(makeRefreshTokenRequest())) {
+        respBodyString = EntityUtils.toString(httpResponse.getEntity());
+
+        if (httpResponse.getStatusLine().getStatusCode() == HttpStatusCodes.STATUS_CODE_OK) {
+          JsonObject respBody = JsonParser.parseString(respBodyString).getAsJsonObject();
+
+          if (respBody.has(ACCESS_TOKEN) && respBody.has(EXPIRES_IN)) {
+            String newAccessToken = respBody.get(ACCESS_TOKEN).toString().replaceAll("^\"|\"$", "");
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/OAuthManager.java:168]: What is this replace targeting?

File: src/main/java/net/snowflake/ingest/connection/RequestBuilder.java [hunk 15]
```diff
@@ -25,10 +25,7 @@
 import net.snowflake.client.jdbc.internal.apache.http.entity.ContentType;
 import net.snowflake.client.jdbc.internal.apache.http.entity.StringEntity;
 import net.snowflake.client.jdbc.internal.apache.http.impl.client.CloseableHttpClient;
-import net.snowflake.ingest.utils.ErrorCode;
-import net.snowflake.ingest.utils.SFException;
-import net.snowflake.ingest.utils.SnowflakeURL;
-import net.snowflake.ingest.utils.StagedFileWrapper;
+import net.snowflake.ingest.utils.*;
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/RequestBuilder.java:28]: No wildcard imports.

File: src/main/java/net/snowflake/ingest/connection/RequestBuilder.java [hunk 16]
```diff
@@ -61,6 +58,8 @@ public class RequestBuilder {
   // Reference to the telemetry service
   private final TelemetryService telemetryService;
 
+  public final String authType;
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/RequestBuilder.java:61]: Nit, docstring.

File: src/main/java/net/snowflake/ingest/connection/RequestBuilder.java [hunk 17]
```diff
@@ -279,6 +290,22 @@ public RequestBuilder(
     this.host = hostName;
     this.userAgentSuffix = userAgentSuffix;
 
+    // create our security/token manager
+    if (credential instanceof KeyPair) {
+      securityManager =
+          new JWTManager(accountName, userName, (KeyPair) credential, telemetryService);
+      authType = Constants.JWT;
+    } else {
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/RequestBuilder.java:298]: Maybe do a switch statement here and throw an exception if not an instance of Keypair or OAuth.

File: src/main/java/net/snowflake/ingest/connection/RequestBuilder.java [hunk 18]
```diff
@@ -563,6 +590,23 @@ private URI makeGetClientURI(UUID requestId, String pipe) throws URISyntaxExcept
...
+      return new URIBuilder()
+          .setScheme(scheme)
+          .setHost(host)
+          .setPort(port)
+          .setPath(TOKEN_REQUEST_ENDPOINT)
+          .build();
+    } catch (URISyntaxException e) {
+      throw new SFException(e, ErrorCode.MAKE_URI_FAILURE, "Make OAuth token request fail");
+    }
+  }
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/connection/RequestBuilder.java:608]: Could this be in the OAuth manager and move the `TOKEN_REQUEST_ENDPOINT` constant in there?

File: src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtils.java [hunk 19]
```diff
@@ -1,7 +1,6 @@
 package net.snowflake.ingest.streaming.internal;
 
-import static net.snowflake.ingest.utils.Constants.MAX_STREAMING_INGEST_API_CHANNEL_RETRY;
-import static net.snowflake.ingest.utils.Constants.RESPONSE_ERR_GENERAL_EXCEPTION_RETRY_REQUEST;
+import static net.snowflake.ingest.utils.Constants.*;
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtils.java:3]: Nit, no wildcards

File: src/main/java/net/snowflake/ingest/connection/JWTManager.java (hunk 1 above)
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/JWTManager.java]: Fixed

File: src/main/java/net/snowflake/ingest/connection/RequestBuilder.java (hunk 18 above)
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/RequestBuilder.java]: Fixed

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java (hunk 14 above)
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/OAuthManager.java:165]: The parsed result would be "<token>", this regex is trimming the start and end quotation.

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java (hunk 11 above)
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/OAuthManager.java]: I implemented at [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/49dd6a8a8749eca4d4927df377524ff1553d7e85/src/main/java/net/snowflake/ingest/connection/OAuthManager.java#L156)

**Comment by sfc-gh-xhuang** [comment]
@sfc-gh-alhuang Is this PR still needed or did your other PRs cover it? 
If it's not needed, please close

**Pull request closed without merging**